src/rgw/rgw_sal_motr.cc

```diff
@@ -1923,127 +1922,130 @@ int MotrObject::MotrDeleteOp::delete_obj(const DoutPrefixProvider* dpp, optional
-        std::string delete_marker_key = source->get_key().to_str();
-        ldpp_dout(dpp, 20)<<__func__<< "Add delete marker in bucket index, key=  " <<  delete_marker_key << dendl;
-        rc = source->store->do_idx_op_by_name(bucket_index_iname,
-                                              M0_IC_PUT, delete_marker_key, del_mark_bl);
-        if(rc < 0)
-        {
-          ldpp_dout(dpp, 0) << "Failed to add delete marker in bucket." << dendl;
+        result.version_id = source->get_instance();
+        // update is-latest=false for current version entry.
+        ldpp_dout(dpp, 20)<<__func__<< " Updating previous version entries " << dendl;
+        rc = source->update_version_entries(dpp);
+        if (rc < 0)
           return rc;
         }
+      // creating a delete marker
+      bufferlist del_mark_bl;
+      rgw_bucket_dir_entry ent_del_marker;
+      ent_del_marker.key.name = source->get_name();
+      ent_del_marker.key.instance = source->get_instance();
+      if(ent.key.instance == "null")
+        ent_del_marker.key.instance = "null";
+      ent_del_marker.meta.owner = params.obj_owner.get_id().to_str();
+      ent_del_marker.meta.owner_display_name = params.obj_owner.get_display_name();
+      ent_del_marker.flags = rgw_bucket_dir_entry::FLAG_DELETE_MARKER | rgw_bucket_dir_entry::FLAG_CURRENT;
+      if (real_clock::is_zero(params.mtime))
+        ent_del_marker.meta.mtime = real_clock::now();
+      else
+        ent_del_marker.meta.mtime = params.mtime;
 
-        // Update in the cache.
-        source->store->get_obj_meta_cache()->put(dpp, delete_marker_key, del_mark_bl);
-      }
-    }else
-    {
-      // Unversioned flow
-      if(ent.meta.size !=0 )
+      rgw::sal::Attrs attrs;
+      ent_del_marker.encode(del_mark_bl);
+      encode(attrs, del_mark_bl);
+      ent_del_marker.meta.encode(del_mark_bl);
+
+      // key for delete marker - obj1[delete-markers's ver-id].
+      std::string delete_marker_key = source->get_key().to_str();
+      ldpp_dout(dpp, 20)<<__func__<< "Add delete marker in bucket index, key=  " <<  delete_marker_key << dendl;
+      rc = source->store->do_idx_op_by_name(bucket_index_iname,
+                                            M0_IC_PUT, delete_marker_key, del_mark_bl);
+      if(rc < 0)
       {
-        if (ent.meta.category == RGWObjCategory::MultiMeta)
-          rc = source->delete_part_objs(dpp);
-        else
-          rc = source->delete_mobj(dpp);
-        if (rc < 0) {
-          ldpp_dout(dpp, 0) << "Failed to delete the object from Motr. " << dendl;
-          return rc;
-        }
+        ldpp_dout(dpp, 0) << "Failed to add delete marker in bucket." << dendl;
+        return rc;
       }
-      rc = source->store->do_idx_op_by_name(bucket_index_iname,
-                                                  M0_IC_DEL, delete_key, bl);
+
+      // Update in the cache.
+      source->store->get_obj_meta_cache()->put(dpp, delete_marker_key, del_mark_bl);
```

Comment:
We should return from here.

Reply:
There are some common code written in the same function which need to be executed after this block. Hence we are not returning from here.

Reply:
Here we are only deleting the delete marker, and common code if for updating the stats which is not required in case delete marker.

Reply:
okay sure